### ISSUE TITLE
Extend use-package to :ensure packages are installed via el-get

### ISSUE THREAD

**Issue by jonEbird** [dimitri/el-get#2156]
I would like to start using John Wiegley's use-package and after reading about its `:ensure` keyword to actually perform the package installation, I wanted to add support for using el-get to perform the installation. My [PR](https://github.com/jwiegley/use-package/pull/200) was basically a dup of the :ensure keyword but works to invoke an `el-get-install`. 

Wiegley suggested that a separate module be built that extends use-package to use el-get and I'm now wondering if that would be appropriate to add directly here? If you're okay with that idea, could you recommend how to approach adding? (_Looking for advice before attempting to code up the solution_)

**Comment by dimitri** [comment]
Can you write up a quick summary of how `use-package` compares to `el-get-bundle`?

**Comment by jonEbird** [comment]
The motivation for me to use `use-package` is for the convenient deferring of loading packages until you actually use them. I think of it as a nice macro around using `eval-after-load` but includes additional features to bind keys, etc. In the following example, the `color-moccur` is not installed until I use the feature by invoking one of the keybinds:
    
    (use-package color-moccur
      :commands (isearch-moccur isearch-all)
      :bind ("M-s O" . moccur)
      :init
      (bind-key "M-o" 'isearch-moccur isearch-mode-map)
      (bind-key "M-O" 'isearch-moccur-all isearch-mode-map)
      :config
      (use-package moccur-edit)
      :el-get color-moccur)

That is how it worked with my small patch to `use-package`. With that example, the next time I restarted Emacs I wouldn't need to install the package and it would stay unloaded until I invoked one of the key bindings again.

**Comment by DarwinAwardWinner** [comment]
I think the most practical way to do this would be a macro that expands to the code for adding a recipe to `el-get-sources` with `:before (use-package ...)`, followed by calls to `el-get-install` and `el-get-init` for that package.

So for example

```lisp
(use-package-el-get PKGNAME
 ;; el-get-stuff for installation
 :description "DESC"
 :type github
 :pkgname "USER/PKGNAME"
 ;; use-package stuff for initialization
 :init ...
 :bind ...
 :interpreter ...
 :mode ...)
```

Would be transformed into something like

```lisp
(progn
  (add-to-list
   'el-get-sources 
   '(:name PKGNAME
     :description "DESC"
     :type github
     :pkgname "USER/PKGNAME"
     :before
     (use-package PKGNAME
       :init ...
       :bind ...
       :interpreter ...
       :mode ...)))
  (el-get-install PKGNAME)
  (el-get-init PKGNAME))
```

With such a macro, you will have the full generality of el-get in defining new packages from scratch in a DRY fashion (whereas in your pull request, you would need separate code to add the needed recipes to `el-get-sources`). Of course, in the simple case where el-get already has a recipe available, the macro would output `(:name PKGNAME :before (use-package PKGNAME ...))`. But you'll probably have to handle weird edge cases like the el-get package name differing from the feature that needs to be required in `use-package` and such. It might be easier to just perform the above transformation by hand in your own config, rather than writing a module for it.

**Comment by npostavs** [comment]
> In the following example, the color-moccur is not installed until I use the feature by invoking one of the keybinds:

Actually, doesn't the installation happen at macroexpand time, so right on startup (or whenever you compile your init.el)? If the package is already installed then it wouldn't be *loaded* until used.

---

> Wiegley suggested that a separate module be built that extends use-package to use el-get and I'm now wondering if that would be appropriate to add directly here? If you're okay with that idea, could you recommend how to approach adding? (Looking for advice before attempting to code up the solution)

You could add a `use-package-el-get-module.rcp` with `:depends use-package` and modify`use-package-keywords` in the `:post-init` form. The recipe could also take care of installing the code (if it ends up in the el-get repo it could just become a `:type builtin` recipe).

---

> Can you write up a quick summary of how use-package compares to el-get-bundle?

`use-package` provides syntactic sugar for setting key bindings, `diminish`ing mode lighters, adding `auto-mode-alist` entries, and running config before and after (uses `eval-after-load`) loading. `el-get-bundle` has syntactic sugar for setting some of the el-get recipe properties, (e.g. `:name foo/bar` -> `:name bar :pkgname "foo/bar"`), and running config code before and after loading via el-get's `:after` property rather than `eval-after-load` directly* (if the el-get recipe has `:lazy t` that is basically equivalent to `eval-after-load`). 

*Although it seems @tarao has in mind actually calling `(with-)eval-after-load` explicitly, according to what he wrote in the manual, hmm.

**Comment by jonEbird** [comment]
@DarwinAwardWinner I like the idea, though the idea of building the uber-macro myself is a bit daunting. It also appears I was misinformed about how both `el-get-bundle` or `el-get-install` work in nature. I thought they ended up loading the packages they installed but upon re-testing that is not the case. 

@npostavs I don't think the installation occurs at macro expansion time. When I was playing around with the initial code, I could see el-get installing `color-moccur` only after I tried to invoke it.

Your idea of creating a separate recipe for `use-package` sounds like a good idea. I may play with that solution until we come up with something better.

I also was unaware of the `:lazy t` option as I've been a minimal user of el-get up to this point. I may need to read up on the manual.

**Comment by npostavs** [comment]
> It also appears I was misinformed about how both el-get-bundle or el-get-install work in nature. I thought they ended up loading the packages they installed but upon re-testing that is not the case.

> I also was unaware of the :lazy t option as I've been a minimal user of el-get up to this point. I may need to read up on the manual.

It depends on the recipe, if it has `:features` or the `:prepare`/`:post-init` invokes an autoloaded function, loading will be triggered. Using `:lazy t` suppresses `:features` and moves `:post-init` (but not `:prepare`) to after loading time. See also, `el-get-is-lazy`.

> I don't think the installation occurs at macro expansion time. When I was playing around with the initial code, I could see el-get installing color-moccur only after I tried to invoke it.

Hmm, that's different from `:ensure` then, though I don't see anything in the code to make the difference.

**Comment by jonEbird** [comment]
If a package being loaded depends on the recipe, I'd think for the sake of this feature request, I'd like to avoid even evaluating the recipe until needed. Or would setting `el-get-is-lazy` to `t` effectively defer any installation / loading of a package? (_Sorry, I haven't done my el-get homework yet_)

I actually didn't test using the default `:ensure` keyword but instead went directly to trying to create an `:el-get` keyword but as you've noticed the code was nearly identical. I'd still expect `:ensure` to not trigger an installation at macro expansion time but still when the feature is triggered and loaded. 

#### To recap my goals with this feature request:
1. Use `use-package` to help defer package loading / evaluating.
2. Continue to use `el-get` for my package management needs.
3. Ideally keep the configuration for any package concisely in one location. 
  - As opposed to how I currently track my `el-get` packages in a separate config. 
  - Would like to leave special el-get package details in their respective recipe files.

#### Alternatives:
1. Create a custom `use-package` recipe that adds my desired `:el-get` keyword.
2. (_Similar to 1._) Create a separate `el-get-use-package.el` package in a PR that people can require that provides the same thing.
2. Create a new macro that melds `use-package` and something like `el-get-bundle` together.

I think I'm leaning towards alternative 1 still.

**Comment by DarwinAwardWinner** [comment]
I still think the quickest and simplest solution is to write a simple macro that takes in a plist that is a hybrid of el-get and use-package keywords, and which spits out an el-get recipe with a `use-package` form in its `:before` slot. Then you can just add that recipe to `el-get-sources` and proceed as usual with el-get. This way you avoid having to muck about in the internals of either package, for the most part.

**Comment by npostavs** [comment]
> Or would setting el-get-is-lazy to t effectively defer any installation / loading of a package?

To defer installation, avoid calling `el-get-install` (or `el-get`) with the package as argument.

`el-get-is-lazy` will defer loading unless the recipe does something it shouldn't in the `:prepare` phase (please report a bug if you find such a thing), or it has `:lazy nil` (only `cedet.rcp` appears to have that currently, and it probably shouldn't). *Most* recipes will defer loading anyway (PRs to make eager recipes into deferred ones are welcome).

> I'd still expect :ensure to not trigger an installation at macro expansion time but still when the feature is triggered and loaded. 

That's not how it works right now, see jwiegley/use-package#190

**Comment by jonEbird** [comment]
@npostavs You're right, it happens during macroexpand. I was naively confusing evaluation with macroexpand. I just tested by explicitly calling `macroexpand` as in one of the examples in the link you provided.

**Comment by DarwinAwardWinner** [comment]
To be fair, I'd argue that use-package having the installation happen at macro expansion is a bug.

**Comment by jonEbird** [comment]
@DarwinAwardWinner Sorry but creating a new macro that combines the plists still seems daunting for my elisp skill level. 

Just working out the elisp part of a custom use-package recipe that would define a `:el-get` keyword for me and define the necessary functions. Just need to now work it into a recipe but until now I've only created simple el-get recipes. Really going to read the info pages this time. 

https://gist.github.com/jonEbird/1e6746893b9059d1408e

**Comment by priyadarshan** [comment]
From my humble (still avid) `el-get` user point of view, it would be very nice to add `use-package` features to `el-get-bundle`.  Perhaps in the future @tarao will consider it? I would love to depend only on `el-get` and drop `use-package`.

Perhaps the `bind-key` library could be somehow relied upon (https://github.com/jwiegley/use-package/blob/master/bind-key.el).

**Comment by priyadarshan** [comment]
Having a look at the (v2) [use-package](https://github.com/jwiegley/use-package) docs, it seems it can be extended with custom handlers and it shouldn't be too hard to add a handler similar to :ensure for el-get.

Apparently, [quelpa](https://github.com/quelpa/quelpa-use-package) added a similar feature.

Perhaps it would not be too difficult to adapt their code, to add a new handler, :el-get, so that

```
(use-package abc-mode :el-get)
```

Would be the equivalent of

```
(el-get-bundle abc-mode)
(use-package abc-mode :ensure t)
```

**Comment by priyadarshan** [comment]
I was still chasing for a solution for #2156, when I found [req-package](https://github.com/edvorg/req-package#el-get), from README, "There is another benefit over use-package - el-get support. No more thinking about sources for your packages. Just install and configure your el-get."
